p4-16/psa/PSA.mdk

```diff
@@ -172,18 +172,21 @@ backend is responsible for mapping the PRE extern APIs to the
 appropriate mechanisms for performing these operations in the
 hardware.
 
-The PRE is instantiated by the architecture and a P4 program can use
-it directly. It is an error to instantiate the PRE multiple times.
+The PRE extern object has no constructor, and thus it cannot be
+instantiated in the user's P4 program. The architecture instantiates
+it exactly once, without requiring the user's P4 program to
+instantiate it.
 The PRE is made available to the Ingress programmable block using the
 same mechanism as packet_in. A corresponding Buffering and Queuing
 Engine (BQE) extern is defined for the Egress pipeline (@see
 BufferingQueueingEngine).
 
-Note: some of these operations may not be implemented as primitive
-operations on a certain target. However, All of the operations can be
-implemented as a combination of other operations. Applications that
-rely on non-primitive operations may incur significant performance
-penalties, however, they should be functionally correct.
+**Note to implementers**: some of these operations may not be
+implemented as primitive operations on a certain target. However, the
+goal is that all operations should be implementable as a combination
+of other operations. Target documentation should highlight such
```

Comment:
"may not be implemented as primitive operations" - What is a primitive operation?  Such things were defined in the P4_14 spec, but the word "primitive" does not occur in the P4_16 v1.0.0 spec, not even once.

"should be implementable as a combination of other operations."  Of what other operations?  Of other operations that could have been written as P4_16 assignment statements?  Of other PRE extern method calls?

The second one doesn't make sense, because the implementation must "bottom out" somewhere.

If the intended meaning is the first, "P4_16 assignment statements", what is the importance of stating this?  The implementer could implement it this way if their chosen implementation is such that P4_16 assignment statements can achieve the desired internal state modification, but it is really up to them.  And the cost of some P4_16 assignment statements can be way more expensive than others, in a target-dependent way, so not sure you are really giving users of the PSA any kind of reassurance with a statement like this.